File: src/entropic/core/worktree.py
"""Git worktree management for delegation isolation.

Branch model:
  main     — pristine, only updated on explicit acceptance
  develop  — working branch, created from main at first delegation
  delegation/* — child branches off develop, merge back to develop

Creates isolated git worktrees for child delegation loops so that
filesystem operations in child contexts don't affect the parent's
working tree. Merges changes back to develop on success, discards on
failure. Develop merges to main only on explicit acceptance.
"""

from __future__ import annotations

import asyncio
import shutil
from contextlib import asynccontextmanager
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Any

from entropic.core.logging import get_logger

if TYPE_CHECKING:
    from collections.abc import AsyncIterator

    from entropic.core.engine import AgentEngine

logger = get_logger("core.worktree")

# Directory under repo root where worktrees are created
WORKTREE_DIR = ".worktrees"

# Branch names
DEVELOP_BRANCH = "develop"


async def _run_git(cwd: Path, args: list[str]) -> tuple[bool, str]:
    """Run a git command, return (success, output)."""
    process = await asyncio.create_subprocess_exec(
        "git",
        *args,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        cwd=cwd,
    )
    stdout, stderr = await process.communicate()
    output = stdout.decode("utf-8", errors="replace").strip()
    error = stderr.decode("utf-8", errors="replace").strip()
    success = process.returncode == 0
    if not success:
        logger.warning("[WORKTREE] git %s failed: %s", " ".join(args), error)
    return success, output or error


async def _current_branch(cwd: Path) -> str:
    """Get the current branch name."""
    _, output = await _run_git(cwd, ["rev-parse", "--abbrev-ref", "HEAD"])
    return output


@dataclass
class WorktreeInfo:
    """Tracks a created worktree's path and branch for later merge/discard."""

    path: Path
    branch: str
    delegation_id: str


class WorktreeManager:
    """Create, merge, and discard git worktrees for delegation isolation.

    Ensures a ``develop`` branch exists before creating delegation
    worktrees. The main working tree is checked out to ``develop``
    so that lead operates there. Delegation branches are created from
    and merged back to ``develop``. Main is only updated via
    ``accept_to_main()``.
    """

    def __init__(self, repo_dir: Path) -> None:
        self._repo_dir = repo_dir
        self._worktree_base = repo_dir / WORKTREE_DIR
        self._develop_ready = False

    async def ensure_develop(self) -> bool:
        """Create develop branch from current HEAD and check it out.

        Idempotent — if develop exists and is checked out, this is a no-op.
        Returns True if the working tree is on develop.
        """
        if self._develop_ready:
            return True

        current = await _current_branch(self._repo_dir)
        if current == DEVELOP_BRANCH:
            self._develop_ready = True
            return True

        ready = await self._create_and_checkout_develop(current)
        self._develop_ready = ready
        return ready

    async def _create_and_checkout_develop(self, current: str) -> bool:
        """Create develop branch if missing and check it out."""
        exists, _ = await _run_git(self._repo_dir, ["rev-parse", "--verify", DEVELOP_BRANCH])
        if not exists:
            success, output = await _run_git(self._repo_dir, ["branch", DEVELOP_BRANCH])
            if not success:
                logger.error("[WORKTREE] Failed to create develop branch: %s", output)
                return False
            logger.info("[WORKTREE] Created develop branch from %s", current)

        success, output = await _run_git(self._repo_dir, ["checkout", DEVELOP_BRANCH])
        if not success:
            logger.error("[WORKTREE] Failed to checkout develop: %s", output)
            return False

        logger.info("[WORKTREE] Working tree now on develop")
        return True

    async def create_worktree(self, delegation_id: str, tier: str) -> WorktreeInfo | None:
        """Create a git worktree for a delegation, branching from develop.

        Returns WorktreeInfo with path and branch, or None on failure.
        """
        if not await self.ensure_develop():
            logger.error("[WORKTREE] Cannot create worktree: develop branch not ready")
            return None

        short_id = delegation_id[:8]
        branch_name = f"delegation/{tier}-{short_id}"
        worktree_path = self._worktree_base / f"delegation-{short_id}"

        self._worktree_base.mkdir(parents=True, exist_ok=True)

        success, output = await _run_git(
            self._repo_dir,
            ["worktree", "add", str(worktree_path), "-b", branch_name, DEVELOP_BRANCH],
        )

        if not success:
            logger.error("[WORKTREE] Failed to create worktree: %s", output)
            return None

        logger.info(
            "[WORKTREE] Created worktree: %s (branch: %s from develop)",
            worktree_path,
            branch_name,
        )
        return WorktreeInfo(path=worktree_path, branch=branch_name, delegation_id=delegation_id)

    async def merge_worktree(self, info: WorktreeInfo) -> bool:
        """Merge worktree branch back to develop.

        The main working tree must be on develop (ensured by create_worktree).
        Returns True if merge succeeded (or nothing to merge).
        """
        await self._auto_commit_if_dirty(info)

        success, output = await _run_git(
            self._repo_dir,
            ["merge", info.branch, "--no-edit"],
        )

        if not success:
            logger.error("[WORKTREE] Merge to develop failed for %s: %s", info.branch, output)
            await self._cleanup(info)
            return False

        logger.info("[WORKTREE] Merged %s into develop", info.branch)
        await self._cleanup(info)
        return True

    async def accept_to_main(self) -> bool:
        """Merge develop into main. Called on explicit acceptance.

        Switches to main, merges develop, switches back to develop.
        Returns True on success.
        """
        # Find the main branch name (main or master)
        main_branch = await self._find_main_branch()
        if main_branch is None:
            logger.error("[WORKTREE] Cannot find main/master branch")
            return False

        success, output = await _run_git(self._repo_dir, ["checkout", main_branch])
        if not success:
            logger.error("[WORKTREE] Failed to checkout %s: %s", main_branch, output)
            return False

        success, output = await _run_git(self._repo_dir, ["merge", DEVELOP_BRANCH, "--no-edit"])
        merge_ok = success
        if not success:
            logger.error("[WORKTREE] Merge develop → %s failed: %s", main_branch, output)

        # Always return to develop
        await _run_git(self._repo_dir, ["checkout", DEVELOP_BRANCH])

        if merge_ok:
            logger.info("[WORKTREE] Accepted: develop merged into %s", main_branch)
        return merge_ok

    async def discard_develop(self) -> bool:
        """Discard develop branch, returning to main. Resets session state.

        Called when the user rejects all delegation work.
        """
        main_branch = await self._find_main_branch()
        if main_branch is None:
            return False

        await _run_git(self._repo_dir, ["checkout", main_branch])
        await _run_git(self._repo_dir, ["branch", "-D", DEVELOP_BRANCH])
        self._develop_ready = False
        logger.info("[WORKTREE] Discarded develop branch, back on %s", main_branch)
        return True

    async def _find_main_branch(self) -> str | None:
        """Find the main branch name (main or master)."""
        for name in ("main", "master"):
            success, _ = await _run_git(self._repo_dir, ["rev-parse", "--verify", name])
            if success:
                return name
        return None

    async def discard_worktree(self, info: WorktreeInfo) -> None:
        """Remove worktree and delete branch without merging."""
        await self._cleanup(info)
        logger.info(
            "[WORKTREE] Discarded worktree %s (branch: %s)",
            info.path.name,
            info.branch,
        )

    async def _auto_commit_if_dirty(self, info: WorktreeInfo) -> None:
        """Auto-commit uncommitted changes in worktree before merge.

        Uses ``git status --porcelain`` so that both modified tracked files
        AND new untracked files are detected.  The previous ``git diff HEAD``
        approach silently missed untracked files, causing child work products
        to be lost on merge.
        """
        if not info.path.exists():
            logger.warning("[WORKTREE] Worktree path missing: %s", info.path)
            return

        success, status = await _run_git(info.path, ["status", "--porcelain"])
        if not success or not status.strip():
            return

        logger.info("[WORKTREE] Auto-committing dirty worktree: %s", info.path.name)
        await _run_git(info.path, ["add", "-A"])
        await _run_git(
            info.path,
            ["commit", "-m", f"delegation/{info.delegation_id[:8]}: child work"],
        )

    async def _cleanup(self, info: WorktreeInfo) -> None:
        """Remove worktree directory and delete branch."""
        if info.path.exists():
            success, _ = await _run_git(
                self._repo_dir,
                ["worktree", "remove", str(info.path), "--force"],
            )
            if not success:
                logger.warning(
                    "[WORKTREE] git worktree remove failed, falling back to rmtree: %s",
                    info.path,
                )
                shutil.rmtree(info.path, ignore_errors=True)

        # Delete the branch
        success, output = await _run_git(self._repo_dir, ["branch", "-D", info.branch])
        if success:
            logger.info("[WORKTREE] Deleted branch %s", info.branch)
        else:
            logger.warning("[WORKTREE] Failed to delete branch %s: %s", info.branch, output)


def _get_server_instance(engine: AgentEngine, name: str) -> Any:
    """Get underlying server instance from engine's server manager."""
    sm = engine.server_manager
    if sm is None:
        return None
    client = sm._clients.get(name)
    if client is None:
        return None
    return getattr(client, "_server", None)


def _swap_git_repo_dir(server: Any, new_dir: Path) -> None:
    """Update repo_dir on GitServer and all its registered tools."""
    server.repo_dir = new_dir
    for tool in server._tool_registry._tools.values():
        if hasattr(tool, "_repo_dir"):
            tool._repo_dir = new_dir


@asynccontextmanager
async def scoped_worktree(
    engine: AgentEngine,
    worktree_path: Path,
) -> AsyncIterator[None]:
    """Swap server directories to worktree, restore on exit.

    Swaps filesystem root_dir, bash working_dir, and git repo_dir
    to the worktree root.  Worktrees are full repo checkouts, so the
    worktree root is the correct working directory — it contains all
    tracked content.  Restores all on exit (success or failure).
    """
    saved: dict[str, Any] = {}

    fs_server = _get_server_instance(engine, "filesystem")
    if fs_server is not None:
        saved["fs_root"] = fs_server.root_dir
        fs_server.root_dir = worktree_path

    bash_server = _get_server_instance(engine, "bash")
    if bash_server is not None:
        saved["bash_dir"] = bash_server.working_dir
        bash_server.working_dir = worktree_path

    git_server = _get_server_instance(engine, "git")
    if git_server is not None:
        saved["git_dir"] = git_server.repo_dir
        _swap_git_repo_dir(git_server, worktree_path)

    try:
        yield
    finally:
        if fs_server is not None and "fs_root" in saved:
            fs_server.root_dir = saved["fs_root"]

        if bash_server is not None and "bash_dir" in saved:
            bash_server.working_dir = saved["bash_dir"]

        if git_server is not None and "git_dir" in saved:
            _swap_git_repo_dir(git_server, saved["git_dir"])

File: src/core/worktree.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later
/**
 * @file worktree.cpp
 * @brief WorktreeManager and ScopedWorktree implementation.
 *
 * Git operations shell out via popen() (matching Python behavior).
 * Each run_git() call captures stdout and returns {success, output}.
 *
 * @version 1.8.6
 */

#include <entropic/core/worktree.h>
#include <entropic/types/logging.h>

#include <array>
#include <cstdio>
#include <random>
#include <set>
#include <sstream>

#include <signal.h>
#include <sys/types.h>
#include <unistd.h>

static auto logger = entropic::log::get("core.worktree");

namespace entropic {

// ── run_git ──────────────────────────────────────────────

/**
 * @brief Run a git command in a directory and capture output.
 * @param repo_dir Working directory for the command.
 * @param args Git subcommand and arguments.
 * @return GitResult with success flag and output.
 * @utility
 * @version 1.8.6
 */
GitResult run_git(const std::filesystem::path& repo_dir,
                  const std::string& args) {
    std::string cmd = "cd '" + repo_dir.string() +
                      "' && git " + args + " 2>&1";

    GitResult result;
    std::array<char, 256> buf{};

    FILE* pipe = popen(cmd.c_str(), "r");
    if (pipe == nullptr) {
        logger->error("popen failed for: git {}", args);
        return result;
    }

    while (fgets(buf.data(), static_cast<int>(buf.size()), pipe) != nullptr) {
        result.output += buf.data();
    }

    int status = pclose(pipe);
    result.success = (status == 0);
    return result;
}

// ── WorktreeManager ──────────────────────────────────────

/**
 * @brief Generate a session id of the form "<pid>-<hex8>".
 *
 * pid is sufficient on its own for liveness detection, but adding
 * random hex avoids collisions when a stale dir lingers after a
 * fast pid-recycle.
 *
 * @return Session identifier string.
 * @utility
 * @version 2.0.6
 */
static std::string make_session_id() {
    std::random_device rd;
    std::uniform_int_distribution<uint32_t> dist;
    char hex[9];
    std::snprintf(hex, sizeof(hex), "%08x", dist(rd));
    return std::to_string(static_cast<long>(getpid())) + "-" + hex;
}

/**
 * @brief Construct with repository root directory.
 *
 * Creates a session-scoped worktree base at
 * `.worktrees/<session_id>/` and prunes stale dirs from prior
 * sessions.
 *
 * @param repo_dir Path to the git repository root.
 * @internal
 * @version 2.0.6
 */
WorktreeManager::WorktreeManager(const std::filesystem::path& repo_dir)
    : repo_dir_(repo_dir),
      session_id_(make_session_id()),
      worktree_base_(repo_dir / ".worktrees" / session_id_) {
    logger->info("WorktreeManager: repo={} session={}",
                 repo_dir_.string(), session_id_);
    prune_stale_worktrees();
    std::error_code ec;
    std::filesystem::create_directories(worktree_base_, ec);
    if (ec) {
        logger->error("Failed to create worktree base {}: {}",
                      worktree_base_.string(), ec.message());
    }
}

/**
 * @brief Destructor — remove this session's worktree base directory.
 *
 * Complements per-worktree cleanup; any delegation directory that
 * leaked past `cleanup()` is removed when the session ends.
 *
 * @internal
 * @version 2.0.6
 */
WorktreeManager::~WorktreeManager() {
    std::error_code ec;
    std::filesystem::remove_all(worktree_base_, ec);
    if (ec) {
        logger->warn("Failed to remove session worktree base {}: {}",
                     worktree_base_.string(), ec.message());
    } else {
        logger->info("Removed session worktree base: {}",
                     worktree_base_.string());
    }
}

/**
 * @brief Check whether the given process id is still alive.
 *
 * Uses `kill(pid, 0)`. Returns false only when errno == ESRCH
 * (owning process is gone). EPERM (different UID) counts as alive —
 * we do not own that session's dir and must leave it.
 *
 * @param pid Process id to probe.
 * @return true if alive or not ours; false if the owner is gone.
 * @utility
 * @version 2.0.6
 */
static bool pid_is_alive(long pid) {
    if (kill(static_cast<pid_t>(pid), 0) == 0) { return true; }
    return errno != ESRCH;
}

/**
 * @brief Remove a directory, logging the outcome.
 * @param dir Directory to remove.
 * @param reason Human-readable reason (for log line).
 * @utility
 * @version 2.0.6
 */
static void remove_orphan_dir(
    const std::filesystem::path& dir, const std::string& reason) {
    std::error_code ec;
    std::filesystem::remove_all(dir, ec);
    if (ec) {
        logger->warn("Failed to remove orphan {} ({}): {}",
                     dir.string(), reason, ec.message());
    } else {
        logger->info("Pruned orphan {} ({})", dir.string(), reason);
    }
}

/**
 * @brief Parse "<pid>-<hex>" session-id directory name into a pid.
 * @param name Directory basename.
 * @param[out] pid Extracted pid on success.
 * @return true if `name` matches the session-id pattern.
 * @utility
 * @version 2.0.6
 */
static bool parse_session_pid(const std::string& name, long& pid) {
    auto dash = name.find('-');
    if (dash == std::string::npos || dash == 0) { return false; }
    try {
        size_t consumed = 0;
        pid = std::stol(name.substr(0, dash), &consumed);
        return consumed == dash && name.size() - dash - 1 == 8;
    } catch (const std::exception&) {
        return false;
    }
}

/**
 * @brief Collect delegation branch names currently in git.
 * @return Set of branch names, trimmed and filtered to delegation/*.
 * @utility
 * @version 2.0.6
 */
std::set<std::string> WorktreeManager::collect_delegation_branches() const {
    std::set<std::string> result;
    auto branches = run_git(repo_dir_, "branch --list 'delegation/*'");
    if (!branches.success || branches.output.empty()) { return result; }

    std::istringstream iss(branches.output);
    std::string line;
    while (std::getline(iss, line)) {
        auto start = line.find_first_not_of(" *+");
        if (start == std::string::npos) { continue; }
        std::string branch = line.substr(start);
        auto end = branch.find_first_of(" \t");
        if (end != std::string::npos) { branch = branch.substr(0, end); }
        if (branch.rfind("delegation/", 0) == 0) {
            result.insert(std::move(branch));
        }
    }
    return result;
}

/**
 * @brief Prune one entry in .worktrees/ if it is a recognizable orphan.
 * @param entry Directory entry under .worktrees/.
 * @utility
 * @version 2.0.6
 */
void WorktreeManager::prune_worktrees_entry(
    const std::filesystem::directory_entry& entry) {
    if (!entry.is_directory()) { return; }
    std::string name = entry.path().filename().string();
    if (name == session_id_) { return; } // our own

    long pid = 0;
    if (parse_session_pid(name, pid)) {
        if (!pid_is_alive(pid)) {
            remove_orphan_dir(entry.path(),
                              "dead session pid=" + std::to_string(pid));
        }
        return;
    }

    // Pre-v2.0.6 layout: `.worktrees/delegation-<id>/` sat directly
    // under .worktrees. Anything matching that prefix here is an
    // orphan from before the session-scoped layout.
    if (name.rfind("delegation-", 0) == 0) {
        remove_orphan_dir(entry.path(), "pre-v2.0.6 layout");
    }
}

/**
 * @brief Remove stale delegation worktrees and branches from prior sessions.
 *
 * Called at startup to clean up any worktrees/branches left behind by
 * crashed or interrupted sessions. Without this, `worktree add -b` fails
 * with "branch already exists" and blocks all delegation.
 *
 * Walk order:
 * 1. `git worktree prune` — sync git admin state with filesystem.
 * 2. Iterate `.worktrees/*`: remove old-layout and dead-session dirs
 *    that git doesn't recognize as worktrees.
 * 3. Delete `delegation/*` branches, after pruning their backing dirs.
 *
 * @internal
 * @version 2.0.6
 */
void WorktreeManager::prune_stale_worktrees() {
    run_git(repo_dir_, "worktree prune");

    auto base = repo_dir_ / ".worktrees";
    std::error_code ec;
    if (std::filesystem::exists(base, ec)) {
        for (const auto& entry :
             std::filesystem::directory_iterator(base, ec)) {
            prune_worktrees_entry(entry);
        }
    }

    auto branches = collect_delegation_branches();
    for (const auto& branch : branches) {
        run_git(repo_dir_, "branch -D " + branch);
    }
    if (!branches.empty()) {
        logger->info("Pruned {} stale delegation branch(es)",
                     branches.size());
    }
}

/**
 * @brief Get the repository root directory.
 * @return Repo root path.
 * @internal
 * @version 1.8.6
 */
const std::filesystem::path& WorktreeManager::repo_dir() const {
    return repo_dir_;
}

/**
 * @brief Find the main branch name (main or master).
 * @return Branch name, or empty string if not found.
 * @internal
 * @version 1.8.6
 */
std::string WorktreeManager::find_main_branch() const {
    auto r = run_git(repo_dir_, "branch --list main");
    if (r.success && r.output.find("main") != std::string::npos) {
        return "main";
    }
    r = run_git(repo_dir_, "branch --list master");
    if (r.success && r.output.find("master") != std::string::npos) {
        return "master";
    }
    return "";
}

/**
 * @brief Create develop branch from current HEAD and check it out.
 * @return true if the working tree is on develop.
 * @internal
 * @version 1.8.6
 */
bool WorktreeManager::ensure_develop() {
    if (develop_ready_) {
        return true;
    }

    auto check = run_git(repo_dir_, "branch --list develop");
    bool exists = check.output.find("develop") != std::string::npos;
    std::string cmd = exists ? "checkout develop" : "checkout -b develop";

    auto result = run_git(repo_dir_, cmd);
    if (!result.success) {
        logger->error("Failed to setup develop branch");
        return false;
    }

    if (!exists) {
        logger->info("Created develop branch from HEAD");
    }
    develop_ready_ = true;
    return true;
}

/**
 * @brief Create a git worktree for a delegation.
 * @param delegation_id Short ID for the delegation.
 * @param tier Target tier name.
 * @return WorktreeInfo on success, nullopt on failure.
 * @internal
 * @version 2.0.6
 */
std::optional<WorktreeInfo> WorktreeManager::create_worktree(
    const std::string& delegation_id,
    const std::string& tier) {
    if (!develop_ready_) {
        logger->error("Cannot create worktree: develop not ready");
        return std::nullopt;
    }

    std::string short_id = delegation_id.substr(
        0, std::min(delegation_id.size(), size_t{8}));
    std::string branch = "delegation/" + tier + "-" + short_id;
    auto path = worktree_base_ / ("delegation-" + short_id);

    std::string cmd = "worktree add -b " + branch +
                      " '" + path.string() + "' develop";
    auto r = run_git(repo_dir_, cmd);
    if (!r.success) {
        logger->warn("create_worktree failed, cleaning stale state: {}",
                     r.output);
        run_git(repo_dir_, "worktree remove --force '"
                           + path.string() + "'");
        std::error_code ec;
        std::filesystem::remove_all(path, ec);
        run_git(repo_dir_, "branch -D " + branch);
        r = run_git(repo_dir_, cmd);
        if (!r.success) {
            logger->error("create_worktree failed after cleanup: {}",
                          r.output);
            return std::nullopt;
        }
    }

    logger->info("Created worktree: {} -> {}", branch, path.string());
    return WorktreeInfo{path, branch, short_id};
}

/**
 * @brief Auto-commit uncommitted changes before merge.
 * @param info Worktree to check.
 * @internal
 * @version 1.8.6
 */
void WorktreeManager::auto_commit_if_dirty(const WorktreeInfo& info) {
    auto status = run_git(info.path, "status --porcelain");
    if (!status.success || status.output.empty()) {
        return;
    }

    run_git(info.path, "add -A");
    run_git(info.path,
            "commit -m 'Auto-commit before merge (" +
            info.delegation_id + ")'");
    logger->info("Auto-committed dirty worktree: {}", info.branch);
}

/**
 * @brief Merge worktree branch back to develop.
 * @param info Worktree to merge.
 * @return true if merge succeeded.
 * @internal
 * @version 1.8.6
 */
bool WorktreeManager::merge_worktree(const WorktreeInfo& info) {
    auto_commit_if_dirty(info);

    auto merge = run_git(repo_dir_,
                         "merge --no-ff " + info.branch +
                         " -m 'Merge delegation " +
                         info.delegation_id + "'");
    bool ok = merge.success;

    if (!ok) {
        logger->error("Merge failed for {}: {}", info.branch, merge.output);
    } else {
        logger->info("Merged {} into develop", info.branch);
    }

    cleanup(info);
    return ok;
}

/**
 * @brief Remove worktree directory and delete branch.
 * @param info Worktree to clean up.
 * @internal
 * @version 1.8.6
 */
/**
 * @brief Remove worktree directory and delete branch.
 *
 * Uses `git worktree remove --force` first, then
 * `std::filesystem::remove_all` as a belt-and-suspenders fallback —
 * if git fails silently (locked fd, permissions), the directory is
 * still removed by the filesystem call.
 *
 * @param info Worktree to clean up.
 * @internal
 * @version 2.0.6
 */
void WorktreeManager::cleanup(const WorktreeInfo& info) {
    run_git(repo_dir_, "worktree remove --force '" +
                       info.path.string() + "'");
    std::error_code ec;
    std::filesystem::remove_all(info.path, ec);
    run_git(repo_dir_, "branch -D " + info.branch);
    logger->info("Cleaned up worktree: {}", info.branch);
}

/**
 * @brief Remove worktree and delete branch without merging.
 * @param info Worktree to discard.
 * @internal
 * @version 1.8.6
 */
void WorktreeManager::discard_worktree(const WorktreeInfo& info) {
    logger->info("Discarding worktree: {}", info.branch);
    cleanup(info);
}

/**
 * @brief Merge develop into main.
 * @return true if merge succeeded.
 * @internal
 * @version 1.8.6
 */
bool WorktreeManager::accept_to_main() {
    std::string main_branch = find_main_branch();
    if (main_branch.empty()) {
        logger->error("No main/master branch found");
        return false;
    }

    auto co = run_git(repo_dir_, "checkout " + main_branch);
    if (!co.success) {
        return false;
    }

    auto merge = run_git(repo_dir_, "merge --no-ff develop "
                                    "-m 'Accept develop into " +
                                    main_branch + "'");
    run_git(repo_dir_, "checkout develop");

    if (!merge.success) {
        logger->error("accept_to_main failed: {}", merge.output);
    }
    return merge.success;
}

/**
 * @brief Discard develop branch, returning to main.
 * @return true if discard succeeded.
 * @internal
 * @version 1.8.6
 */
bool WorktreeManager::discard_develop() {
    std::string main_branch = find_main_branch();
    if (main_branch.empty()) {
        return false;
    }

    auto co = run_git(repo_dir_, "checkout " + main_branch);
    if (!co.success) {
        return false;
    }

    auto del = run_git(repo_dir_, "branch -D develop");
    develop_ready_ = false;
    return del.success;
}

// ── ScopedWorktree ───────────────────────────────────────

/**
 * @brief Construct and swap directories.
 * @param swap_fn Callback to swap all server directories.
 * @param user_data Opaque pointer for swap_fn.
 * @param worktree_path Target worktree directory.
 * @param original_path Original working directory to restore.
 * @internal
 * @version 1.8.6
 */
ScopedWorktree::ScopedWorktree(
    SwapDirFn swap_fn,
    void* user_data,
    const std::filesystem::path& worktree_path,
    const std::filesystem::path& original_path)
    : swap_fn_(swap_fn),
      user_data_(user_data),
      original_path_(original_path) {
    if (swap_fn_ != nullptr) {
        swap_fn_(worktree_path, user_data_);
        logger->info("Swapped dirs to: {}", worktree_path.string());
    }
}

/**
 * @brief Restore original directories.
 * @internal
 * @version 1.8.6
 */
ScopedWorktree::~ScopedWorktree() {
    if (swap_fn_ != nullptr) {
        swap_fn_(original_path_, user_data_);
        logger->info("Restored dirs to: {}", original_path_.string());
    }
}

} // namespace entropic
